File: backend/app/schemas/episode.py
"""HTTP contracts for Neşeli Orman episode generation."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class EpisodeGenerateRequest(BaseModel):
    """Validated input accepted by the episode generation endpoint."""

    theme_id: str = Field(min_length=1, max_length=100)
    project_id: uuid.UUID | None = None


class EpisodeBatchGenerateRequest(BaseModel):
    """Validated input accepted by the batch episode generation endpoint."""

    project_id: uuid.UUID | None = None


class BatchGeneratedEpisodeItem(BaseModel):
    """One episode created by a batch-generate call."""

    id: uuid.UUID
    theme_id: str
    theme_label: str
    title: str


class EpisodeBatchGenerateResponse(BaseModel):
    """Result of generating every theme's episode for one project in one call."""

    project_id: uuid.UUID | None
    created: list[BatchGeneratedEpisodeItem]
    skipped_theme_ids: list[str]


class ThemeSummaryResponse(BaseModel):
    """A theme's id, Turkish label, and cast/location images, for a theme picker."""

    theme_id: str
    label: str
    lead_character_name: str
    lead_character_image_url: str
    lead_character_voice_sample_url: str
    support_character_name: str
    support_character_image_url: str
    support_character_voice_sample_url: str
    location_name: str
    location_image_url: str
    location_ambient_video_url: str


class EpisodeResponse(BaseModel):
    """Transport-safe representation of a generated episode script."""

    theme_id: str
    theme_label: str
    title: str
    lead_character: dict[str, Any]
    support_character: dict[str, Any]
    location: dict[str, Any]
    lesson: str
    total_duration_seconds: int
    scenes: list[dict[str, Any]]


class SeoPackageResponse(BaseModel):
    """Transport-safe representation of a generated SEO package."""

    titles: list[str]
    description: str
    tags: list[str]
    thumbnail_suggestion: str


class ShortsPlanResponse(BaseModel):
    """Transport-safe representation of a generated Shorts cut plan."""

    episode_theme_id: str
    total_duration_seconds: int
    segments: list[dict[str, Any]]


class EpisodeGenerationResponse(BaseModel):
    """Combined output of the episode, SEO, and Shorts generation services."""

    id: uuid.UUID
    project_id: uuid.UUID | None
    episode: EpisodeResponse
    seo: SeoPackageResponse
    shorts: ShortsPlanResponse


class GeneratedEpisodeSummaryResponse(BaseModel):
    """One row of the persisted-episode history list."""

    id: uuid.UUID
    project_id: uuid.UUID | None
    title: str
    theme_id: str
    theme_label: str
    lead_character_name: str
    lead_character_image_url: str
    lead_character_voice_sample_url: str
    support_character_name: str
    support_character_image_url: str
    support_character_voice_sample_url: str
    location_name: str
    location_image_url: str
    location_ambient_video_url: str
    created_at: datetime


class GeneratedEpisodeListResponse(BaseModel):
    """One newest-first page of persisted episode summaries."""

    items: list[GeneratedEpisodeSummaryResponse]
    total: int
    page: int
    page_size: int

File: frontend/src/types/episode.ts
export type ThemeSummary = {
  theme_id: string
  label: string
  lead_character_image_url: string
  support_character_image_url: string
  location_image_url: string
}

export type VoiceProfile = {
  pitch: string
  pace: string
  tone: string
  catchphrase: string
}

export type EpisodeCharacter = {
  character_id: string
  name: string
  species: string
  role: string
  core_value: string
  personality: string[]
  visual_description: string
  voice: VoiceProfile
  image_url: string
}

export type EpisodeLocation = {
  location_id: string
  name: string
  type: string
  description: string
  image_url: string
}

export type EpisodeScene = {
  name: string
  duration_seconds: number
  text: string
  speaker: string | null
  dialogue: string | null
}

export type Episode = {
  theme_id: string
  theme_label: string
  title: string
  lead_character: EpisodeCharacter
  support_character: EpisodeCharacter
  location: EpisodeLocation
  lesson: string
  total_duration_seconds: number
  scenes: EpisodeScene[]
}

export type SeoPackage = {
  titles: string[]
  description: string
  tags: string[]
  thumbnail_suggestion: string
}

export type ShortsSegment = {
  name: string
  duration_seconds: number
  text: string
}

export type ShortsPlan = {
  episode_theme_id: string
  total_duration_seconds: number
  segments: ShortsSegment[]
}

export type EpisodeGenerationResult = {
  id: string
  project_id: string | null
  episode: Episode
  seo: SeoPackage
  shorts: ShortsPlan
}

export type GeneratedEpisodeSummary = {
  id: string
  project_id: string | null
  title: string
  theme_id: string
  theme_label: string
  lead_character_image_url: string
  support_character_image_url: string
  location_image_url: string
  created_at: string
}

export type GeneratedEpisodeList = {
  items: GeneratedEpisodeSummary[]
  total: number
  page: number
  page_size: number
}
